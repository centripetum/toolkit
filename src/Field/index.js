import * as React from 'react'

import { isEmpty, not } from 'ramda'

import getButton from '../helpers/getButton'
import getInput from '../helpers/getInput'
import getLabel from '../helpers/getLabel'
import useFieldState from '../hooks/useFieldState'

export default function Field (props) {
  const {
    clearButton,
    defaultValue,
    handleChange,
    input,
    label,
    labelText,
    name,
    noClear,
    noLabel,
    noReset,
    noUndoRedo,
    onSave,
    redoButton,
    resetButton,
    saveButton,
    type,
    undoButton
  } = props

  const {
    clear,
    dirty,
    noRedo,
    noUndo,
    onChange,
    redo,
    reset,
    undo,
    value
  } = useFieldState({
    defaultValue,
    handleChange,
    noClear,
    noReset,
    noUndoRedo
  })
  const inputType = type || 'text'
  const handleSave = onSave ? () => onSave(name, value) : null

  return (
    <>
      {getLabel(noLabel, label, name, labelText)}
      {getInput(input, name, onChange, inputType, value)}
      {getButton(undoButton, noUndo, 'Undo', undo)}
      {getButton(redoButton, noRedo, 'Redo', redo)}
      {getButton(clearButton, isEmpty(value), 'Clear', clear)}
      {getButton(saveButton, value === defaultValue, 'Save', handleSave)}
      {getButton(resetButton, not(dirty), 'Reset', reset)}
    </>
  )
}
