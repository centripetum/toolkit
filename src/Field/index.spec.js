import * as React from 'react'
import { render } from 'react-testing-library'

import Field from './'

describe('Field', () => {
  it('works', () => {
    const { container } = render(
      <Field
        defaultValue='Everything and log'
        labelText='Message'
        name='message'
        onSave={() => null}
      />
    )

    expect(container).toMatchSnapshot()
  })
})
