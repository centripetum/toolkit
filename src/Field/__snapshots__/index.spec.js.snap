// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field works 1`] = `
<div>
  <label
    class="sc-htpNat iKBgbH"
    for="message"
  >
    Message
  </label>
  <input
    class="sc-bwzfXH lfkgtm"
    name="message"
    type="text"
    value="Everything and log"
  />
  <button
    class="sc-bdVaJa dxWZKA"
    disabled=""
  >
    Undo
  </button>
  <button
    class="sc-bdVaJa dxWZKA"
    disabled=""
  >
    Redo
  </button>
  <button
    class="sc-bdVaJa jgAgQk"
  >
    Clear
  </button>
  <button
    class="sc-bdVaJa dxWZKA"
    disabled=""
  >
    Save
  </button>
  <button
    class="sc-bdVaJa dxWZKA"
    disabled=""
  >
    Reset
  </button>
</div>
`;
